### Like v3.2.1 BUT with 0.05 test split.

In [1]:
from keras.engine import Model
from keras.layers import Flatten, Dense, Input, Dropout
from keras_vggface.vggface import VGGFace
from keras.preprocessing.image import ImageDataGenerator
from keras_vggface import utils
from keras.optimizers import Adam
from keras.preprocessing import image
from keras import backend as K
from keras.callbacks import ModelCheckpoint, EarlyStopping
from sklearn.utils import class_weight
import keras
import numpy as np
import os
import shutil
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
%matplotlib inline

/home/gloria/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/home/gloria/tensorflow/lib/python3.6/importlib/_bootstrap.py:205: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


In [2]:
# Constants
IMG_SIZE = 200
BASE_DIR = '../project/all_females'

In [3]:
# Read in ratings
ratings = np.genfromtxt('./Projekt_SGE_Assessment_ErikK.txt')

In [4]:
# Round (obviously imbalanced sets)
ratings = np.round(ratings, 0).astype(np.int8)
print(np.unique(ratings, return_counts=True))

(array([0, 1, 2, 3, 4, 5], dtype=int8), array([197, 128,  75,  70,  50,  48]))


In [5]:
### Distribute images across folders
for r in np.unique(ratings).astype(np.int8):
    if not os.path.isdir('./data/train/{}-stars'.format(r)):
        os.makedirs('./data/train/{}-stars'.format(r))
    if not os.path.isdir('./data/test/{}-stars'.format(r)):
        os.makedirs('./data/test/{}-stars'.format(r))

In [6]:
### Move all images according to ratings
images = sorted(os.listdir(BASE_DIR))

### Make sure lengths of ratings and images correspond 
assert len(ratings) == len(images)

### Sort images according to ratings
ratings = ratings.tolist()
ratings, images = zip(*sorted(zip(ratings,images)))
ratings = np.asarray(ratings)

### Count number of ratings for each category, generate train test split for each category
unique, counts = np.unique(ratings, return_counts=True)

stars = [[] for i in counts]
stared_img = [[] for i in counts]
accum = np.cumsum(counts)
stars[0] = ratings[:accum[0]]
stared_img[0] = images[:accum[0]]

for i in range(1,len(counts)):
    stars[i] = ratings[accum[i-1]:accum[i]]
    stared_img[i] = images[accum[i-1]:accum[i]]

stars = np.asarray(stars)
stared_img = np.asarray(stared_img)

assert(stars.shape == stared_img.shape)

X_train = [[] for i in counts]
X_test = [[] for i in counts]
y_train = [[] for i in counts]
y_test = [[] for i in counts]

for i in range(len(counts)):
    X_train[i], X_test[i], y_train[i], y_test[i] = train_test_split(stared_img[i], stars[i], test_size=0.2)
    
X_train = np.asarray(X_train)
X_test = np.asarray(X_test)
y_train = np.asarray(y_train)
y_test = np.asarray(y_test)

        
print("Train-test-split:")
print("X_train: ",X_train.shape)
print("X_test: ",X_test.shape)
print("y_train: ",y_train.shape)
print("y_test: ",y_test.shape)

#check that all subsets of the split have the same rating
for i in range(len(counts)):
    assert(np.sum(y_train[i]) == i*len(y_train[i]))
    assert(np.sum(y_test[i]) == i*len(y_test[i]))

Train-test-split:
X_train:  (6,)
X_test:  (6,)
y_train:  (6,)
y_test:  (6,)


In [7]:
### Loop throgh all images and store them in the dedicated folders
#for i, r in enumerate(ratings.astype(np.int8)):
#    src = os.path.join(BASE_DIR, images[i])
#    dest = os.path.join('./classification/{}-stars'.format(r), images[i])
#    shutil.copy(src, dest)

for i in range(len(X_train)):
    for j in range(len(X_train[i])):
        src = os.path.join(BASE_DIR, X_train[i][j])
        dest = os.path.join('./data/train/{}-stars'.format(y_train[i][j]), X_train[i][j])
        shutil.copy(src, dest)

for i in range(len(X_test)):
    for j in range(len(X_test[i])):
        src = os.path.join(BASE_DIR, X_test[i][j])
        dest = os.path.join('./data/test/{}-stars'.format(y_test[i][j]), X_test[i][j])
        shutil.copy(src, dest)

## Time to use data augmentation with streaming

In [8]:
datagen = ImageDataGenerator(rescale=1./255, 
                             rotation_range=40,
                             width_shift_range=0.2,
                             height_shift_range=0.2,
                             shear_range=0.2,
                             zoom_range=0.2,
                             horizontal_flip=True,
                             fill_mode='nearest',
                             preprocessing_function=utils.preprocess_input)

In [9]:
train_generator = datagen.flow_from_directory(
    './data/train/',
    target_size=(IMG_SIZE, IMG_SIZE),
    batch_size=64,
    class_mode='categorical'#, save_to_dir='aug_data/train', save_prefix='aug_train', save_format='png'
)

test_generator = datagen.flow_from_directory(
    './data/test/',
    target_size=(IMG_SIZE, IMG_SIZE),
    batch_size=64,
    class_mode='categorical'#, save_to_dir='aug_data/test', save_prefix='aug_test', save_format='png'
)

Found 548 images belonging to 6 classes.
Found 210 images belonging to 6 classes.


## Some random model

In [10]:
# Add custom layers
vgg_base = VGGFace(include_top=False, input_shape=(IMG_SIZE, IMG_SIZE, 3), pooling='max')
last_layer = vgg_base.get_layer('global_max_pooling2d_1').output
#X = Dropout(0.2)(last_layer)
X = Dense(256, activation='relu', name='fc6')(last_layer)
X = Dropout(0.2)(X)
X = Dense(128, activation='relu', name='fc7')(X)
X = Dropout(0.2)(X)
X = Dense(64, activation='relu', name='fc8')(X)
X = Dropout(0.2)(X)
output = Dense(len(np.unique(ratings)), activation='softmax')(X)
model = Model(inputs=vgg_base.input, outputs=output)

In [11]:
# Freeze vgg layers
for layer in vgg_base.layers:
    layer.trainable = False

In [12]:
# Compute class weights since we have unbalaned classes
class_weight = class_weight.compute_class_weight('balanced', 
                                                 np.unique(ratings), 
                                                 ratings)

In [13]:
model.compile(optimizer=Adam(clipnorm=1.0),
              loss='categorical_crossentropy', 
              metrics=['accuracy'])

In [14]:
checkpointer = ModelCheckpoint(filepath='./weights-v3.2.1.4.hdf5', verbose=1, save_best_only=True)
earlystop = EarlyStopping(monitor='val_loss', patience=10, verbose=0)

In [15]:
history = model.fit_generator(
    train_generator,
    steps_per_epoch=100,
    epochs=1,
    class_weight=class_weight,
    validation_data=test_generator,
    callbacks=[checkpointer]
)

Epoch 1/1
100/100 [==============================] - 73s 731ms/step - loss: 1.5959 - acc: 0.3523 - val_loss: 1.4200 - val_acc: 0.4571

Epoch 00001: val_loss improved from inf to 1.42002, saving model to ./weights-v3.2.1.4.hdf5


In [16]:
history = model.fit_generator(
    train_generator,
    steps_per_epoch=453//64,
    epochs=500,
    class_weight=class_weight,
    validation_data=test_generator,
    callbacks=[earlystop, checkpointer]
)

Epoch 1/500
7/7 [==============================] - 8s 1s/step - loss: 1.4490 - acc: 0.4286 - val_loss: 1.4279 - val_acc: 0.4190

Epoch 00001: val_loss did not improve
Epoch 2/500
7/7 [==============================] - 6s 920ms/step - loss: 1.5223 - acc: 0.3939 - val_loss: 1.4287 - val_acc: 0.4476

Epoch 00002: val_loss did not improve
Epoch 3/500
7/7 [==============================] - 7s 969ms/step - loss: 1.4557 - acc: 0.4018 - val_loss: 1.3908 - val_acc: 0.4524

Epoch 00003: val_loss improved from 1.42002 to 1.39075, saving model to ./weights-v3.2.1.4.hdf5
Epoch 4/500
7/7 [==============================] - 6s 928ms/step - loss: 1.4164 - acc: 0.4159 - val_loss: 1.3900 - val_acc: 0.4476

Epoch 00004: val_loss improved from 1.39075 to 1.38999, saving model to ./weights-v3.2.1.4.hdf5
Epoch 5/500
7/7 [==============================] - 7s 973ms/step - loss: 1.4497 - acc: 0.3705 - val_loss: 1.3675 - val_acc: 0.4333

Epoch 00005: val_loss improved from 1.38999 to 1.36752, saving model to ./w

Epoch 43/500
7/7 [==============================] - 7s 941ms/step - loss: 1.2005 - acc: 0.5067 - val_loss: 1.1108 - val_acc: 0.5857

Epoch 00043: val_loss did not improve
Epoch 44/500
7/7 [==============================] - 6s 919ms/step - loss: 1.1640 - acc: 0.5158 - val_loss: 1.1628 - val_acc: 0.5476

Epoch 00044: val_loss did not improve
Epoch 45/500
7/7 [==============================] - 6s 924ms/step - loss: 1.1381 - acc: 0.5585 - val_loss: 1.0653 - val_acc: 0.5857

Epoch 00045: val_loss did not improve
Epoch 46/500
7/7 [==============================] - 6s 926ms/step - loss: 1.1081 - acc: 0.5668 - val_loss: 1.0842 - val_acc: 0.6048

Epoch 00046: val_loss did not improve
Epoch 47/500
7/7 [==============================] - 7s 953ms/step - loss: 1.1165 - acc: 0.5536 - val_loss: 1.1172 - val_acc: 0.5524

Epoch 00047: val_loss did not improve
Epoch 48/500
7/7 [==============================] - 6s 922ms/step - loss: 1.1566 - acc: 0.5352 - val_loss: 1.1091 - val_acc: 0.5810

Epoch 00048:

7/7 [==============================] - 7s 956ms/step - loss: 0.7642 - acc: 0.7254 - val_loss: 0.7049 - val_acc: 0.7429

Epoch 00086: val_loss improved from 0.71970 to 0.70491, saving model to ./weights-v3.2.1.4.hdf5
Epoch 87/500
7/7 [==============================] - 7s 950ms/step - loss: 0.7972 - acc: 0.7143 - val_loss: 0.7005 - val_acc: 0.7714

Epoch 00087: val_loss improved from 0.70491 to 0.70048, saving model to ./weights-v3.2.1.4.hdf5
Epoch 88/500
7/7 [==============================] - 7s 943ms/step - loss: 0.7891 - acc: 0.7079 - val_loss: 0.7538 - val_acc: 0.7762

Epoch 00088: val_loss did not improve
Epoch 89/500
7/7 [==============================] - 6s 923ms/step - loss: 0.8006 - acc: 0.7057 - val_loss: 0.7676 - val_acc: 0.7190

Epoch 00089: val_loss did not improve
Epoch 90/500
7/7 [==============================] - 6s 927ms/step - loss: 0.7273 - acc: 0.7387 - val_loss: 0.6897 - val_acc: 0.7667

Epoch 00090: val_loss improved from 0.70048 to 0.68968, saving model to ./weight

7/7 [==============================] - 6s 891ms/step - loss: 0.5875 - acc: 0.7636 - val_loss: 0.6356 - val_acc: 0.7905

Epoch 00130: val_loss did not improve
Epoch 131/500
7/7 [==============================] - 7s 960ms/step - loss: 0.4953 - acc: 0.8192 - val_loss: 0.5925 - val_acc: 0.8095

Epoch 00131: val_loss did not improve
Epoch 132/500
7/7 [==============================] - 6s 921ms/step - loss: 0.5719 - acc: 0.7943 - val_loss: 0.5466 - val_acc: 0.8524

Epoch 00132: val_loss did not improve
Epoch 133/500
7/7 [==============================] - 7s 951ms/step - loss: 0.5974 - acc: 0.7812 - val_loss: 0.4674 - val_acc: 0.8714

Epoch 00133: val_loss improved from 0.47173 to 0.46744, saving model to ./weights-v3.2.1.4.hdf5
Epoch 134/500
7/7 [==============================] - 6s 920ms/step - loss: 0.5069 - acc: 0.8071 - val_loss: 0.6871 - val_acc: 0.7905

Epoch 00134: val_loss did not improve
Epoch 135/500
7/7 [==============================] - 6s 924ms/step - loss: 0.5378 - acc: 0.7875 

In [17]:
model.save('./models/love_classifierv3.2.1.4.h5')

### Choose candidates from test set

In [18]:
#freeze layers
#for layer in model.layers:
#    layer.trainable = False

#transform X_test & y_test so all data is in one dimension 
realNumDim = 0
for i in range(X_test.shape[0]):
    for j in range(len(X_test[i])):
        realNumDim += 1
    
#get test data (unmodified)
data_test = np.zeros((realNumDim, IMG_SIZE, IMG_SIZE, 3))    
last_set = 0
for i in range(X_test.shape[0]): #iterate over #stars
    for idx, _im in enumerate(X_test[i]): #iterate over images
        _img = image.load_img(os.path.join(BASE_DIR, _im), target_size=(IMG_SIZE,IMG_SIZE))
        _x = image.img_to_array(_img)
        _x = np.expand_dims(_x, axis=0)
        data_test[last_set, :, :, :] = utils.preprocess_input(_x, version=1) / .255
        last_set += 1

predictions = model.predict(data_test)
pred_classes = predictions.argmax(axis=-1)

print(pred_classes)
print(y_test)

[0 0 0 0 0 0 0 0 0 2 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 3 0 0 0 0 0 0 0 0 0
 0 0 0 1 1 1 3 3 1 1 1 0 1 1 0 1 1 1 1 1 1 2 1 1 1 1 5 1 1 4 4 2 5 2 2 2 2
 2 0 2 2 2 2 2 3 0 3 3 3 3 3 3 3 3 3 3 3 3 0 4 1 1 4 4 4 4 2 4 5 0 5 5 3 5
 5 3 5 5]
[array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
 array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1])
 array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])
 array([3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3])
 array([4, 4, 4, 4, 4, 4, 4, 4, 4, 4])
 array([5, 5, 5, 5, 5, 5, 5, 5, 5, 5])]


### More training data improves val_acc from 0.5000 (see rater-v3.2.1.3) to 0.8714.